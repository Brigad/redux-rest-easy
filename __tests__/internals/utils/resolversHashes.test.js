import {
  computeNewResolversHashes,
  getEmptyResourceHash,
  getPayloadIdsHash,
  getResourceHash,
  getResourcesHash,
  resetResourceResolversHashes,
} from '../../../src/internals/utils/resolversHashes';

const RESOURCE_NAME = 'fruits';
const NORMALIZED_URL = 'eat:https://api.co/fruits';
const PRINCIPAL_RESOURCE_IDS = ['2', '3', '1'];
const NORMALIZED_PAYLOAD = {
  [RESOURCE_NAME]: {
    1: {
      type: 'apple',
      juicy: true,
    },
    2: {
      type: 'banana',
      juicy: false,
    },
    3: {
      type: 'cherry',
      juicy: false,
    },
  },
};
const EMPTY_STATE = {};
const FILLED_STATE = {
  requests: {
    [NORMALIZED_URL]: {
      payloadIds: {
        [RESOURCE_NAME]: PRINCIPAL_RESOURCE_IDS,
      },
    },
  },
  resources: NORMALIZED_PAYLOAD,
};

const EMPTY_STATE_COMPUTED_HASHES = {
  ...EMPTY_STATE,
  resolversHashes: computeNewResolversHashes(
    EMPTY_STATE,
    RESOURCE_NAME,
    NORMALIZED_URL,
    NORMALIZED_PAYLOAD,
    PRINCIPAL_RESOURCE_IDS,
  ),
};
const FILLED_STATE_COMPUTED_HASHES = {
  ...FILLED_STATE,
  resolversHashes: computeNewResolversHashes(
    FILLED_STATE,
    RESOURCE_NAME,
    NORMALIZED_URL,
    NORMALIZED_PAYLOAD,
    PRINCIPAL_RESOURCE_IDS,
  ),
};
const FILLED_STATE_RESET_HASHES = {
  ...FILLED_STATE,
  resolversHashes: resetResourceResolversHashes(FILLED_STATE, RESOURCE_NAME),
};

describe('computeNewResolversHashes', () => {
  test('empty state', () => {
    const hashBeforeComputing = getResourcesHash(EMPTY_STATE);

    const hashAfterComputing = getResourcesHash(
      EMPTY_STATE_COMPUTED_HASHES.resolversHashes,
    );

    expect(hashBeforeComputing).not.toBe(hashAfterComputing);
  });

  test('filled state', () => {
    const hashBeforeComputing = getResourcesHash(FILLED_STATE);

    const hashAfterComputing = getResourcesHash(
      FILLED_STATE_COMPUTED_HASHES.resolversHashes,
    );

    expect(hashBeforeComputing).not.toBe(hashAfterComputing);
  });
});

describe('resetResourceResolversHashes', () => {
  test('only path', () => {
    const hashBeforeComputing = getResourceHash(
      FILLED_STATE.resolversHashes,
      RESOURCE_NAME,
    );

    const hashAfterComputing = getResourceHash(
      FILLED_STATE_COMPUTED_HASHES.resolversHashes,
      RESOURCE_NAME,
    );

    const hashAfterResetting = getResourceHash(
      FILLED_STATE_RESET_HASHES.resolversHashes,
      RESOURCE_NAME,
    );

    expect(hashBeforeComputing).not.toBe(hashAfterComputing);
    expect(hashBeforeComputing).toBe(hashAfterResetting);
  });
});

describe('getEmptyResourceHash', () => {
  test('only path', () => {
    expect(getEmptyResourceHash()).toMatchSnapshot();
  });
});

describe('getPayloadIdsHash', () => {
  test('no resolversHashes', () => {
    expect(getPayloadIdsHash()).toBe(getEmptyResourceHash());
  });

  test('no resolversHashes.requests', () => {
    expect(getPayloadIdsHash({})).toBe(getEmptyResourceHash());
  });

  test('no normalizedURL', () => {
    expect(getPayloadIdsHash(EMPTY_STATE_COMPUTED_HASHES.resolversHashes)).toBe(
      getEmptyResourceHash(),
    );
  });

  test('no resourceName', () => {
    expect(
      getPayloadIdsHash(
        EMPTY_STATE_COMPUTED_HASHES.resolversHashes,
        NORMALIZED_URL,
      ),
    ).toBe(getEmptyResourceHash());
  });

  test('without computing first', () => {
    expect(
      getPayloadIdsHash(
        FILLED_STATE.resolversHashes,
        NORMALIZED_URL,
        RESOURCE_NAME,
      ),
    ).toBe(getEmptyResourceHash());
  });

  test('after computing', () => {
    expect(
      getPayloadIdsHash(
        FILLED_STATE_COMPUTED_HASHES.resolversHashes,
        NORMALIZED_URL,
        RESOURCE_NAME,
      ),
    ).toMatchSnapshot();
  });
});

describe('getResourcesHash', () => {
  test('no resolversHashes', () => {
    expect(getResourcesHash()).toBe(getEmptyResourceHash());
  });

  test('no resolversHashes.requests', () => {
    expect(getResourcesHash({})).toBe(getEmptyResourceHash());
  });

  test('without computing first', () => {
    expect(getResourcesHash(FILLED_STATE.resolversHashes)).toBe(
      getEmptyResourceHash(),
    );
  });

  test('after computing', () => {
    expect(
      getResourcesHash(FILLED_STATE_COMPUTED_HASHES.resolversHashes),
    ).toMatchSnapshot();
  });
});

describe('getResourceHash', () => {
  test('no resolversHashes', () => {
    expect(getResourceHash()).toBe(getEmptyResourceHash());
  });

  test('no resolversHashes.requests', () => {
    expect(getResourceHash({})).toBe(getEmptyResourceHash());
  });

  test('no resourceName', () => {
    expect(getResourceHash(FILLED_STATE_COMPUTED_HASHES.resolversHashes)).toBe(
      getEmptyResourceHash(),
    );
  });

  test('without computing first', () => {
    expect(getResourceHash(FILLED_STATE.resolversHashes, RESOURCE_NAME)).toBe(
      getEmptyResourceHash(),
    );
  });

  test('after computing', () => {
    expect(
      getResourceHash(
        FILLED_STATE_COMPUTED_HASHES.resolversHashes,
        RESOURCE_NAME,
      ),
    ).toMatchSnapshot();
  });
});
