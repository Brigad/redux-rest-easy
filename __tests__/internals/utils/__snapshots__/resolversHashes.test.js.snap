// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEmptyResourceHash only path 1`] = `"cf252fdcd0c577918a558e3d5869780d"`;

exports[`getPayloadIdsHash after computing 1`] = `"f00555c74464ef36bbc004bcf654aca1"`;

exports[`getResourceHash after computing 1`] = `"2d64c2982ff64c3288c39158728ba72c"`;

exports[`getResourcesHash after computing 1`] = `"529477faa63b29b600bab57fb91ba873"`;
